### PR TITLE
configure.ac: check for AC_FUNC_ALLOCA

### PR DESCRIPTION
**Pull request by iamleot**
Add AC_FUNC_ALLOCA and adjust src/i965_decoder_utils.c to properly include
relevant header files for alloca(3).

Some operating systems does not have <alloca.h> (e.g. NetBSD). This fixes
compilation on such systems.

Signed-off-by: Leonardo Taccari <iamleot@gmail.com>

### CODE REVIEW

**Comment by yakuizhao** [comment]
It seems that the alloca is used for the buffer allocation that is automatically freed.
Maybe we can remove the alloca and manage the required buffer manually.
Will you please use the malloc/free instead of alloca and then remove the header file of "alloca.h"?

File: src/i965_decoder_utils.c [hunk 1]
```diff
@@ -23,7 +23,17 @@
 
 #include "sysdeps.h"
 #include <limits.h>
-#include <alloca.h>
+#ifdef STDC_HEADERS
+# include <stdlib.h>
```
Comment (xhaihao) [inline src/i965_decoder_utils.c]: sysdeps.h has already included stdlib.h, it would be better to move this check to sysdeps.h

**Comment by xhaihao** [comment]
@yakuizhao for some cases, alloca is more efficient than malloc/free

**Comment by yakuizhao** [comment]
@xhaihao , sometimes the alloca seems higher efficiency than malloc/free.  But based on the user-manual of alloca, it is machine and compiler-dependent.  Except the longjmp/siglongjmp, the alloca is discouraged.
So I suggest that it can be replaced.

**Comment by xhaihao** [comment]
Okay, It is fine for me to replace alloca with malloc/free. the code is only used for some legacy platforms.

**Comment by iamleot** [comment]
Hello Xiang,

"Xiang, Haihao" writes:
> xhaihao commented on this pull request.
> 
> 
> 
> > @@ -23,7 +23,17 @@
>  
>  #include "sysdeps.h"
>  #include <limits.h>
> -#include <alloca.h>
> +#ifdef STDC_HEADERS
> +# include <stdlib.h>
> 
> sysdeps.h has already included stdlib.h, it would be better to move this check to sysdeps.h
> 

Sorry, I have not noticed it. I will address that (but please see
further comments)!


Thanks for pointing that out!

**Comment by iamleot** [comment]
Hello Xiang,

"Xiang, Haihao" writes:
> Okay, It is fine for me to replace alloca with malloc/free. the code is only used for some legacy platforms.

Allright, I will try to replace alloca(3) with malloc(3)/free(3) as
discussed in the next hours (probably not before 20:00 UTC or so).


Thank you very much for yours and yakuizhao's suggestions about it!

**Comment by iamleot** [comment]
Hello @xhaihao and @yakuizhao,
I have just tried to address it as discussed and tested on NetBSD.

Please let me know if any further adjustments are needed!


Thank you very much again!

PS: Not checking malloc(3) return value should be correct in that case
because that's already handled in dri_bo_get_subdata().

**Comment by xhaihao** [comment]
@iamleot could you squash the two commits into one commit for this issue?

**Comment by iamleot** [comment]
Hello Xiang,

"Xiang, Haihao" writes:
> @iamleot could you squash the two commits into one commit for this issue?
> 

Sure, I have just done it! (that's the first time that I've used
`git rebase -i`, so please let me know if I have m[ei]ssed something!)

File: src/i965_decoder_utils.c [hunk 2]
```diff
@@ -355,6 +354,9 @@ avc_get_first_mb_bit_offset_with_epb(
             i += 2, j++, n++;
     }
 
+    if (buf)
+        free(buf);
+
```
Comment (xhaihao) [inline src/i965_decoder_utils.c]: free() can accept NULL pointer, so it is unnecessary to check the pointer against NULL.

**Comment by iamleot** [comment]
Hello Xiang,

"Xiang, Haihao" writes:
> > @@ -355,6 +354,9 @@ avc_get_first_mb_bit_offset_with_epb(
>              i += 2, j++, n++;
>      }
>  
> +    if (buf)
> +        free(buf);
> +
> 
> free() can accept NULL pointer, so it is unnecessary to check the pointer against NULL.
> 

OK! I have just adjusted it and squashed the various commits again.

**Comment by xhaihao** [comment]
@iamleot Could you rebase your branch against the latest master branch in https://github.com/01org/intel-vaapi-driver.git ? I got conflict when doing rebase.  In addition, I still see a few new commits besides the merge commits  in your branch, could you squash your fixes into a single commit?

**Comment by iamleot** [comment]
Hello Xiang,

"Xiang, Haihao" writes:
> @iamleot Could you rebase your branch against the latest master branch in https://github.com/01org/intel-vaapi-driver.git ? I got conflict when doing rebase.  In addition, I still see a few new commits besides the merge commits  in your branch, could you squash your fixes into a single commit?

OK! I have just rebased to the latest master branch and squashed the
commits again.

**Pull request merged**